addons/l10n_pk_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,470 @@
+                ]
+            }
+        }
+        Description: This response includes multiple error messages in the `errorMessage`
+        field and the `errors["$values"]` array.
+        """
+        if response.get("errorMessage"):
+            error_message = response["errorMessage"]
+            return {
+                invoice: {
+                    "success": False,
+                    "error": error_message,
+                    "blocking_level": "error",
+                }
+            }
+        json_dump = json.dumps(response)
+        json_name = "%s_einvoice.json" % (invoice.name.replace("/", "_"))
+        attachment = self.env["ir.attachment"].create({
+            "name": json_name,
+            "raw": json_dump.encode(),
+            "res_model": "account.move",
+            "res_id": invoice.id,
+            "mimetype": "application/json",
+        })
+        return {
+            invoice: {
+                "success": True,
+                "attachment": attachment
+            }
+        }
+
+    def _get_l10n_pk_total_tax_values(self, tax_lines, move):
+        sales_tax_total = 0.0
+        non_sales_tax_total = 0.0
+        sale_group_ids = (
+            self._get_l10n_pk_sales_groups(move.company_id)
+            + self._get_l10n_pk_withhold_groups(move.company_id)
+        ).ids
+        for line in tax_lines:
+            if line in sale_group_ids:
+                sales_tax_total += tax_lines[line]['amount']
+            else:
+                non_sales_tax_total += tax_lines[line]['amount']
+        return sales_tax_total, non_sales_tax_total
+
+    def _get_l10n_pk_tax_details_by_line(self, tax_details):
+        l10n_pk_tax_details = {}
+        for tax_detail in tax_details.values():
+            line_code = tax_detail["tax"].tax_group_id.id
+            l10n_pk_tax_detail = l10n_pk_tax_details.get(line_code, {})
+            l10n_pk_tax_detail.update(dict.fromkeys(("rate", "amount", "amount_currency"), 0.0))
+            l10n_pk_tax_detail["rate"] += tax_detail["tax"].amount
+            l10n_pk_tax_detail["amount"] += tax_detail["tax_amount"]
+            l10n_pk_tax_detail["amount_currency"] += tax_detail["tax_amount_currency"]
+            l10n_pk_tax_details[line_code] = l10n_pk_tax_detail
+        return l10n_pk_tax_details
+
+    # ================================ API methods ===========================
+
+    @api.model
+    def _l10n_pk_edi_no_config_response(self):
+        return {
+            'statusCode': 0,
+            'errorMessage': _(
+                "A token is still needs to be set or it's wrong for the E-invoice(Pakistan). "
+                "It needs to be added and verify in the Settings."
+            )
+        }
+
+    @api.model
+    def _l10n_pk_edi_get_token(self, company):
+        return company.sudo().l10n_pk_edi_token or False
+
+    @api.model
+    def _l10n_pk_edi_connect_to_server(self, company=None, payload=None, timeout=None):
```

Comment:
```suggestion
    def _l10n_pk_edi_connect_to_server(self, company, payload, timeout):
```
I don't see any of these param as optional